### PR TITLE
[DataStream] Introduce converter for fluss row to json

### LINKED ISSUE #678: Introduce converter for Fluss row to json
### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.

### Description

We can take a look at `RowDataToJsonConverters` of Flink for this

### Willingness to contribute

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by gyang94**
### Purpose

Linked issue: close #678 

### Brief change log

Add `FlussRowToJsonConverters`

### Tests

`FlussDeserializationSchemaTest.testJsonStringDeserialize`

### CODE REVIEW

**Comment by gyang94** [comment]
Hi @polyzos , this PR is not completed, but the main functions are almost there, you can take a look first.
And I have several questions before continue, could you help?
1. There are no `ARRAY`, `MAP` and `ROW` filedGetter in the `InternalRow.createFieldGetter`. But we have `DataTypes.ARRAY`, `DataTypes.MAP` and `DataTypes.ROW`. It is confusing. Can we have array, map or nested row fields in `InternalRow`? If yes, don't we need fieldGetters for them?
2. In flink `RowDataToJsonConverters`, there are some configurations about json parsing. I think we can also add them in your `JsonStringDeserializationSchema`. Is that good? (In this PR, I only add a default timestampFormat value).
>  private final TimestampFormat timestampFormat;
    private final JsonFormatOptions.MapNullKeyMode mapNullKeyMode;
    private final String mapNullKeyLiteral;
    private final boolean ignoreNullFields;
3. Currently this json schema and converter classes are put in the `flink-fluss-common` module. It is not a good place. Suppose we should have a `fluss-formats` module for all ser/der schema related classes?
4. In the `JsonDeserialzationSchema`, you uses the "flink" shaded jackson libs. I think it should be "fluss" shaded jackson, thus I modified it in this PR. Is there any concern? Please correct me if I am wrong.

Thanks in advance. cc: @wuchong

**Comment by wuchong** [comment]
Hi @gyang94 ,

> 1. There are no ARRAY, MAP and ROW filedGetter in the InternalRow.createFieldGetter. But we have DataTypes.ARRAY, DataTypes.MAP and DataTypes.ROW. It is confusing. Can we have array, map or nested row fields in InternalRow? If yes, don't we need fieldGetters for them?

Fluss doesn't support ARRAY, MAP, ROW for now. But the community has a plan to support them. It's not jsut adding fields to `InternalRow`, each of them have very complex work to do. So let's just ignore the confusion for now. 

> 2. In flink RowDataToJsonConverters, there are some configurations about json parsing. I think we can also add them in your JsonStringDeserializationSchema. Is that good? (In this PR, I only add a default timestampFormat value).

+1 for adding more features/options to it. But would be better to support them in another issue and PR to make this PR foucs. Besides, would be better to extend `JsonStringDeserializationSchema` to have a builder class instead of have a lot of constructors. 

> 3. Currently this json schema and converter classes are put in the flink-fluss-common module. It is not a good place. Suppose we should have a fluss-formats module for all ser/der schema related classes?

It's fine to put them the current module for now. We can revisit it when the `fluss-formats` is needed. 

> 4. In the JsonDeserialzationSchema, you uses the "flink" shaded jackson libs. I think it should be "fluss" shaded jackson, thus I modified it in this PR. Is there any concern? Please correct me if I am wrong.

+1

**Comment by gyang94** [comment]
@wuchong Thanks for the answer. Then there is not too much code needs to be changed for now. I will remove the WIP status and this PR is ready to review.

For more json parsing options and builder class, I will add them in another issue.

BTW, comparing to flink row, what makes the difference in Fluss about array, map and row in `InternalRow`? It may be an interesting topic for me.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/source/deserializer/JsonStringDeserializationSchema.java [hunk 1]
```diff
@@ -68,13 +71,33 @@ public class JsonStringDeserializationSchema implements FlussDeserializationSche
...
+    private transient FlussRowToJsonConverters.FlussRowDataToJsonConverter runtimeConverter;
+
+    /** Timestamp format specification which is used to parse timestamp. */
+    private final TimestampFormat timestampFormat;
+
+    public JsonStringDeserializationSchema() {
+        this(TimestampFormat.ISO_8601);
+    }
+
+    public JsonStringDeserializationSchema(TimestampFormat timestampFormat) {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/source/deserializer/JsonStringDeserializationSchema.java]: Change this method to `private` and provide a builder pattern for the class for future extension. We can keep the 0-parameter constructor public for simple cases.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/TimestampFormat.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.fluss.flink.utils;
+
+import com.alibaba.fluss.annotation.Internal;
+
+/** Timestamp format Enums. */
+@Internal
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/TimestampFormat.java]: `@PublicEvolving` as it is exposed in `JsonStringDeserializationSchema`.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/TimeFormats.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.temporal.ChronoField;
+
+/**
+ * Time formats and timestamp formats respecting the RFC3339 specification, ISO-8601 specification
+ * and SQL specification.
+ */
+@Internal
+public class TimeFormats {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/TimeFormats.java]: remove the `public` keyword to keep it package accessible to avoid the class naming confusion with `TimestampFormat` for users.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    public FlussRowToJsonConverters(TimestampFormat timestampFormat) {
+        this.timestampFormat = timestampFormat;
+    }
+
+    /**
+     * Runtime converter that converts objects of Fluss data structures to corresponding {@link
+     * JsonNode}s.
+     */
+    @FunctionalInterface
+    public interface FlussRowDataToJsonConverter extends Serializable {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: `FlussRowToJsonConverter`

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+            int days = (int) value;
+            LocalDate date = LocalDate.ofEpochDay(days);
+            return mapper.getNodeFactory().textNode(ISO_LOCAL_DATE.format(date));
+        });
+    }
+
+    private FlussRowDataToJsonConverter createTimeConverter() {
+        return (mapper, reuse, value) -> {
+            int seconds = (int) value;
+            LocalTime time = LocalTime.ofSecondOfDay(seconds);
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: the value is milli-seconds and thus this should be divided `1000L`.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+                            .textNode(ISO8601_TIMESTAMP_FORMAT.format(timestamp.toLocalDateTime()));
+                };
+            case SQL:
+                return (mapper, reuse, value) -> {
+                    TimestampNtz timestamp = (TimestampNtz) value;
+                    return mapper.getNodeFactory()
+                            .textNode(SQL_TIMESTAMP_FORMAT.format(timestamp.toLocalDateTime()));
+                };
+            default:
+                throw new TableException(
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: throw `UnsupportedOperationException` instead.

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                    TimestampLtz timestampWithLocalZone = (TimestampLtz) value;
+                    return mapper.getNodeFactory()
+                            .textNode(
+                                    SQL_TIMESTAMP_WITH_LOCAL_TIMEZONE_FORMAT.format(
+                                            timestampWithLocalZone
+                                                    .toInstant()
+                                                    .atOffset(ZoneOffset.UTC)));
+                };
+            default:
+                throw new TableException(
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: ditto

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+                    return mapper.getNodeFactory()
+                            .textNode(SQL_TIMESTAMP_FORMAT.format(timestamp.toLocalDateTime()));
+                };
+            default:
+                throw new TableException(
+                        "Unsupported timestamp format. Validator should have checked that.");
+        }
+    }
+
+    private FlussRowDataToJsonConverter createTimestampWithLocalZone() {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: `createTimestampLtzConverter`

File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                node = mapper.createObjectNode();
+            } else {
+                node = (ObjectNode) reuse;
+            }
+            InternalRow row = (InternalRow) value;
+            for (int i = 0; i < fieldCount; i++) {
+                String fieldName = fieldNames[i];
+                try {
+                    Object field = fieldGetters[i].getFieldOrNull(row);
+                    if (field != null) {
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/utils/FlussRowToJsonConverters.java]: remvoe the condition to print null fields by default?

File: fluss-common/src/main/java/com/alibaba/fluss/row/TimestampNtz.java [hunk 10]
```diff
@@ -126,6 +127,18 @@ public LocalDateTime toLocalDateTime() {
         return LocalDateTime.of(localDate, localTime);
     }
 
+    /** Converts this {@link TimestampNtz} object to a {@link Instant}. */
+    public Instant toInstant() {
```
Comment (wuchong) [inline fluss-common/src/main/java/com/alibaba/fluss/row/TimestampNtz.java]: remove, not used.

File: fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java [hunk 11]
```diff
@@ -111,31 +125,92 @@ public void testGetProducedType() {
...
+                    @Override
+                    public UserCodeClassLoader getUserCodeClassLoader() {
+                        return null;
+                    }
+
+                    @Override
+                    public RowType getRowSchema() {
+                        return rowType;
+                    }
+                });
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java]: Use `deserializer.open(new InitializationContextImpl(null, null, rowType));` instead for simplication.

File: fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java [hunk 12]
```diff
@@ -111,31 +125,92 @@ public void testGetProducedType() {
...
         String result = deserializer.deserialize(scanRecord);
 
+        String rowJson =
+                "{\"char\":\"a\",\"string\":\"test value\",\"boolean\":false,\"binary\":\"dGVzdCBiaW5hcnk=\",\"bytes\":\"dGVzdCBieXRlcw==\",\"tinyInt\":1,\"smallInt\":64,\"integer\":1024,\"bigInt\":2048,\"float\":1.1,\"double\":3.14,\"date\":\"2025-04-21\",\"timeWithoutTimeZone\":\"10:00:00\",\"timestampWithoutTimeZone\":\"2025-04-21T10:00:00\",\"timestampWithLocalTimeZone\":\"2025-04-21T10:00:00Z\"}";
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java]: json format the string for better reviewing

```java
                "{"
                        + "\"char\":\"a\","
                        + "\"string\":\"test value\","
                        + "\"boolean\":false,"
                        + "\"binary\":\"dGVzdCBiaW5hcnk=\","
                        + "\"bytes\":\"dGVzdCBieXRlcw==\","
                        + "\"tinyInt\":1,"
                        + "\"smallInt\":64,"
                        + "\"integer\":1024,"
                        + "\"bigInt\":2048,"
                        + "\"float\":1.1,"
                        + "\"double\":3.14,"
                        + "\"date\":\"2025-04-21\","
                        + "\"timeWithoutTimeZone\":\"10:00:00\","
                        + "\"timestampWithoutTimeZone\":\"2025-04-21T10:00:00\","
                        + "\"timestampWithLocalTimeZone\":\"2025-04-21T10:00:00Z\""
                        + "}";
```

File: fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java [hunk 13]
```diff
@@ -111,31 +125,92 @@ public void testGetProducedType() {
...
 
         // Verify with offset and timestamp
         ScanRecord scanRecord2 = new ScanRecord(1001, 1743261788400L, ChangeType.DELETE, row);
         String result2 = deserializer.deserialize(scanRecord2);
         assertThat(result2).isNotNull();
         assertThat(result2)
                 .isEqualTo(
-                        "{\"offset\":1001,\"timestamp\":1743261788400,\"change_type\":\"DELETE\",\"row\":\"(42,test value)\"}");
+                        "{\"offset\":1001,\"timestamp\":1743261788400,\"change_type\":\"DELETE\",\"row\":"
+                                + rowJson
```
Comment (wuchong) [inline fluss-flink/fluss-flink-common/src/test/java/com/alibaba/fluss/flink/source/deserializer/FlussDeserializationSchemaTest.java:214]: 1. add tests for null fields
2. change some fields and verify the result again. (as we are reusing the `ObjectNode` in `JsonStringDeserializationSchema`.)

**Comment by gyang94** [comment]
@wuchong comments addressed, please check.

**Review by wuchong — APPROVED** [review]
